apinetlet/controllers/natgateway_controller.go

```diff
@@ -0,0 +1,277 @@
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/onmetal/controller-utils/clientutils"
+	onmetalapinetv1alpha1 "github.com/onmetal/onmetal-api-net/api/v1alpha1"
+	apinetletv1alpha1 "github.com/onmetal/onmetal-api-net/apinetlet/api/v1alpha1"
+	commonv1alpha1 "github.com/onmetal/onmetal-api/api/common/v1alpha1"
+	networkingv1alpha1 "github.com/onmetal/onmetal-api/api/networking/v1alpha1"
+	"github.com/onmetal/onmetal-api/apiutils/predicates"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	natGatewayFinalizer = "apinet.api.onmetal.de/natgateway"
+)
+
+type NatGatewayReconciler struct {
+	client.Client
+	APINetClient client.Client
+
+	APINetNamespace string
+
+	WatchFilterValue string
+}
+
+//+kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+//+kubebuilder:rbac:groups=networking.api.onmetal.de,resources=natgateways,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=networking.api.onmetal.de,resources=natgateways/finalizers,verbs=update;patch
+//+kubebuilder:rbac:groups=networking.api.onmetal.de,resources=natgateways/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=apinet.api.onmetal.de,resources=publicips,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apinet.api.onmetal.de,resources=publicips/status,verbs=get
+
+func (r *NatGatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	natGateway := &networkingv1alpha1.NATGateway{}
+	if err := r.Get(ctx, req.NamespacedName, natGateway); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("error getting nat gateway %s: %w", req.NamespacedName, err)
+		}
+
+		return r.deleteGone(ctx, log, req.NamespacedName)
+	}
+
+	return r.reconcileExists(ctx, log, natGateway)
+}
+
+func (r *NatGatewayReconciler) deleteGone(ctx context.Context, log logr.Logger, natGatewayKey client.ObjectKey) (ctrl.Result, error) {
+	log.V(1).Info("Delete gone")
+
+	log.V(1).Info("Deleting any matching apinet public ips")
+	if err := r.APINetClient.DeleteAllOf(ctx, &onmetalapinetv1alpha1.PublicIP{},
+		client.InNamespace(r.APINetNamespace),
+		client.MatchingLabels{
+			apinetletv1alpha1.NATGatewayNamespaceLabel: natGatewayKey.Namespace,
+			apinetletv1alpha1.NATGatewayNameLabel:      natGatewayKey.Name,
+		},
+	); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error deleting apinet public ips: %w", err)
+	}
+
+	log.V(1).Info("Issued delete for any leftover apinet public ip")
```

Comment:
Isn't the deletion finished at this point in time? (instead of using the wording "Issued")